File: sdks/go/providers/cartesia/protocol.go
package cartesia

import (
	"bytes"
	"encoding/base64"
	"encoding/json"
	"errors"
	"fmt"
	"math"
	"strconv"
	"strings"

	out "github.com/speechswitch/client/sdks/go/generated/cartesia_output"
	"github.com/speechswitch/client/sdks/go/runtime"
)

type Error struct {
	StatusCode                   float64
	ErrorCode                    *string
	RequestID, DocURL, ContextID runtime.Optional[string]
	title, message               string
}

func (e *Error) Error() string {
	return fmt.Sprintf("Cartesia %s: %s: %s", strconv.FormatFloat(e.StatusCode, 'f', -1, 64), e.title, e.message)
}

func responseError(value any, status float64) *Error {
	result := &Error{StatusCode: status, title: "Request failed", message: "Invalid error response"}
	if text, ok := value.(string); ok {
		result.message = text
	}
	if fields, ok := value.(map[string]any); ok {
		if title, ok := fields["title"].(string); ok {
			result.title = title
		}
		if message, ok := fields["message"].(string); ok {
			result.message = message
		}
		if code, ok := fields["error_code"].(string); ok {
			result.ErrorCode = &code
		}
		if request, ok := fields["request_id"].(string); ok {
			result.RequestID = runtime.Some(request)
		}
		if doc, ok := fields["doc_url"].(string); ok {
			result.DocURL = runtime.Some(doc)
		}
		if context, ok := fields["context_id"].(string); ok {
			result.ContextID = runtime.Some(context)
		}
	}
	return result
}

func load(data []byte) (any, error) {
	if !runtime.ValidUTF8JSON(data) {
		return nil, errors.New("Cartesia returned invalid JSON")
	}
	var value any
	decoder := json.NewDecoder(bytes.NewReader(data))
	decoder.UseNumber()
	if err := decoder.Decode(&value); err != nil {
		return nil, errors.New("Cartesia returned invalid JSON")
	}
	return value, nil
}

func number(value any) (float64, bool) {
	raw, ok := value.(json.Number)
	if !ok {
		return 0, false
	}
	valueNumber, err := strconv.ParseFloat(string(raw), 64)
	return valueNumber, err == nil
}

type packet struct {
	kind           string
	context, group runtime.Optional[string]
	data           string
	timestamps     []out.Timestamp
	failure        *Error
}

func decode(data []byte, websocket bool) (packet, error) {
	result := packet{}
	parsed, err := load(data)
	if err != nil {
		return result, err
	}
	value, ok := parsed.(map[string]any)
	if !ok {
		return result, errors.New("Cartesia returned an invalid frame")
	}
	status, ok := number(value["status_code"])
	if !ok || math.Trunc(status) != status {
		return result, errors.New("Cartesia returned an invalid status_code")
	}
	if context := value["context_id"]; context != nil {
		text, ok := context.(string)
		if !ok {
			return result, errors.New("Cartesia returned an invalid context ID")
		}
		result.context = runtime.Some(text)
	}
	if raw, present := value["flush_id"]; present {
		group, ok := number(raw)
		if !ok || group < 0 || group > 9007199254740991 || math.Trunc(group) != group {
			return result, errors.New("Cartesia returned an invalid flush ID")
		}
		if group == 0 {
			group = 0
		} // Normalize negative zero to the native integer ID.
		result.group = runtime.Some(strconv.FormatFloat(group, 'f', -1, 64))
	}
	result.kind, _ = value["type"].(string)
	if result.kind == "error" {
		result.failure = responseError(value, status)
		return result, nil
	}
	if websocket && !result.context.Present {
		return result, errors.New("Cartesia WebSocket output lacks its context ID")
	}
	done, ok := value["done"].(bool)
	if !ok {
		return result, errors.New("Cartesia returned an invalid completion flag")
	}
	switch result.kind {
	case "done":
		if done {
			return result, nil
		}
	case "flush_done":
		if websocket && value["flush_done"] == true && result.group.Present {
			return result, nil
		}
	case "chunk":
		if text, ok := value["data"].(string); ok {
			result.data = text
			return result, nil
		}
	case "timestamps", "phoneme_timestamps":
		field, labels := "word_timestamps", "words"
		var kind out.TimestampKind = out.TimestampKindAsWord{}
		if result.kind == "phoneme_timestamps" {
			field, labels, kind = "phoneme_timestamps", "phonemes", out.TimestampKindAsPhoneme{}
		}
		raw, present := value[field]
		result.timestamps = []out.Timestamp{}
		if !present && websocket {
			return result, nil
		}
		timing, ok := raw.(map[string]any)
		if !ok {
			return result, errors.New("Cartesia returned incomplete timestamps")
		}
		words, wordsOK := timing[labels].([]any)
		starts, startsOK := timing["start"].([]any)
		ends, endsOK := timing["end"].([]any)
		if !wordsOK || !startsOK || !endsOK || len(words) != len(starts) || len(words) != len(ends) {
			return result, errors.New("Cartesia returned mismatched timestamp arrays")
		}
		for index, value := range words {
			label, labelOK := value.(string)
			start, startOK := number(starts[index])
			end, endOK := number(ends[index])
			if !labelOK || !startOK || !endOK || start < 0 || end < start || math.IsInf(start*1000, 0) || math.IsInf(end*1000, 0) {
				return result, errors.New("Cartesia returned an invalid timestamp")
			}
			result.timestamps = append(result.timestamps, out.Timestamp{Kind: kind, Value: label, StartTimeMs: start * 1000, EndTimeMs: end * 1000})
		}
		return result, nil
	}
	name := fmt.Sprint(value["type"])
	if value["type"] == nil {
		name = "null"
		if _, present := value["type"]; !present {
			name = "undefined"
		}
	}
	return result, fmt.Errorf("Unknown Cartesia event: %s", name)
}

func output(value packet, context string, timed bool) (out.SynthesisItem, error) {
	if value.failure != nil {
		return nil, value.failure
	}
	if value.kind == "done" {
		return nil, nil
	}
	if value.kind == "flush_done" {
		return out.SynthesisItemAsFlush{Value: out.SynthesisItemFlush{CorrelationId: context, InputGroupId: value.group}}, nil
	}
	envelope := out.TimelineOutput{CorrelationId: context, InputGroupId: value.group, Timestamps: value.timestamps}
	if value.kind == "chunk" {
		if strings.ContainsAny(value.data, "\r\n") {
			return nil, errors.New("Cartesia returned invalid base64 audio")
		}
		audio, err := base64.StdEncoding.DecodeString(value.data)
		if err != nil {
			return nil, errors.New("Cartesia returned invalid base64 audio")
		}
		if !timed && !value.group.Present {
			return out.SynthesisItemAsBytes{Value: audio}, nil
		}
		envelope.Audio, envelope.Timestamps = runtime.Some(audio), []out.Timestamp{}
	}
	return out.SynthesisItemAsTimeline{Value: envelope}, nil
}

File: sdks/rust/src/providers/cartesia/protocol.rs
use crate::{base64, generated::cartesia_output as out, http::TransportError, json::Raw};
use std::{collections::BTreeMap, fmt};

#[derive(Debug)]
pub struct Error {
    pub status_code: f64,
    pub error_code: Option<String>,
    pub request_id: Option<String>,
    pub doc_url: Option<String>,
    pub context_id: Option<String>,
    title: String,
    message: String,
}
impl fmt::Display for Error {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "Cartesia {}: {}: {}",
            self.status_code, self.title, self.message
        )
    }
}
impl std::error::Error for Error {}

pub(super) fn failure(message: impl Into<String>) -> TransportError {
    Box::new(std::io::Error::other(message.into()))
}
pub(super) fn response_error(value: &str, status: f64) -> Error {
    let parsed = Raw::parse_exact(value).ok();
    let fields = parsed.and_then(|v| v.object().ok()).unwrap_or_default();
    let string = |key: &str| fields.get(key).and_then(|v| v.string().ok());
    Error {
        status_code: status,
        error_code: string("error_code"),
        request_id: string("request_id"),
        doc_url: string("doc_url"),
        context_id: string("context_id"),
        title: string("title").unwrap_or_else(|| "Request failed".into()),
        message: string("message").unwrap_or_else(|| match parsed {
            Some(value) => value
                .string()
                .unwrap_or_else(|_| "Invalid error response".into()),
            None => value.into(),
        }),
    }
}

pub(super) struct Context {
    pub id: Option<String>,
    pub group: Option<String>,
}
pub(super) enum Packet {
    Chunk(Context, String),
    Timing(Context, Vec<out::Timestamp>),
    Done(Context),
    Flushed(Context, String),
    Failure(Context, Error),
}
impl Packet {
    pub fn context(&self) -> &Context {
        match self {
            Self::Chunk(v, _)
            | Self::Timing(v, _)
            | Self::Done(v)
            | Self::Flushed(v, _)
            | Self::Failure(v, _) => v,
        }
    }
}

pub(super) fn decode(text: &str, websocket: bool) -> Result<Packet, TransportError> {
    let parsed = Raw::parse_exact(text).map_err(|_| failure("Cartesia returned invalid JSON"))?;
    let fields = parsed
        .object()
        .map_err(|_| failure("Cartesia returned an invalid frame"))?;
    let status = fields
        .get("status_code")
        .and_then(|v| v.number().ok())
        .filter(|v| v.is_finite() && v.fract() == 0.0)
        .ok_or_else(|| failure("Cartesia returned an invalid status_code"))?;
    let id = match fields.get("context_id") {
        None => None,
        Some(v) if v.is_null() => None,
        Some(v) => Some(
            v.string()
                .map_err(|_| failure("Cartesia returned an invalid context ID"))?,
        ),
    };
    let group = match fields.get("flush_id") {
        None => None,
        Some(v) => {
            let n = v
                .number()
                .ok()
                .filter(|v| {
                    v.is_finite() && *v >= 0.0 && *v <= 9007199254740991.0 && v.fract() == 0.0
                })
                .ok_or_else(|| failure("Cartesia returned an invalid flush ID"))?;
            Some((n as u64).to_string())
        }
    };
    let context = Context { id, group };
    let kind = fields.get("type").and_then(|v| v.string().ok());
    if kind.as_deref() == Some("error") {
        return Ok(Packet::Failure(context, response_error(text, status)));
    }
    if websocket && context.id.is_none() {
        return Err(failure("Cartesia WebSocket output lacks its context ID"));
    }
    let done = fields
        .get("done")
        .and_then(|v| v.boolean().ok())
        .ok_or_else(|| failure("Cartesia returned an invalid completion flag"))?;
    match kind.as_deref() {
        Some("done") if done => return Ok(Packet::Done(context)),
        Some("flush_done")
            if websocket
                && fields.get("flush_done").and_then(|v| v.boolean().ok()) == Some(true) =>
        {
            if let Some(group) = context.group.clone() {
                return Ok(Packet::Flushed(context, group));
            }
        }
        Some("chunk") => {
            if let Some(data) = fields.get("data").and_then(|v| v.string().ok()) {
                return Ok(Packet::Chunk(context, data));
            }
        }
        Some("timestamps") | Some("phoneme_timestamps") => {
            let phoneme = kind.as_deref() == Some("phoneme_timestamps");
            let field = if phoneme {
                "phoneme_timestamps"
            } else {
                "word_timestamps"
            };
            let timestamps = match fields.get(field) {
                None if websocket => Vec::new(),
                value => {
                    let timing = value
                        .and_then(|v| v.object().ok())
                        .ok_or_else(|| failure("Cartesia returned incomplete timestamps"))?;
                    timestamps(&timing, phoneme)?
                }
            };
            return Ok(Packet::Timing(context, timestamps));
        }
        _ => {}
    }
    let kind = kind.unwrap_or_else(|| fields.get("type").map_or("undefined", |v| v.text()).into());
    Err(failure(format!("Unknown Cartesia event: {kind}")))
}

fn timestamps(
    value: &BTreeMap<String, Raw<'_>>,
    phoneme: bool,
) -> Result<Vec<out::Timestamp>, TransportError> {
    let arrays = [if phoneme { "phonemes" } else { "words" }, "start", "end"]
        .map(|key| value.get(key).and_then(|v| v.array().ok()));
    let [Some(labels), Some(starts), Some(ends)] = arrays else {
        return Err(failure("Cartesia returned mismatched timestamp arrays"));
    };
    if labels.len() != starts.len() || labels.len() != ends.len() {
        return Err(failure("Cartesia returned mismatched timestamp arrays"));
    }
    labels
        .into_iter()
        .zip(starts)
        .zip(ends)
        .map(|((label, start), end)| {
            let value = label
                .string()
                .map_err(|_| failure("Cartesia returned an invalid timestamp"))?;
            let start = start
                .number()
                .map_err(|_| failure("Cartesia returned an invalid timestamp"))?;
            let end = end
                .number()
                .map_err(|_| failure("Cartesia returned an invalid timestamp"))?;
            if !start.is_finite()
                || !end.is_finite()
                || start < 0.0
                || end < start
                || !(start * 1000.0).is_finite()
                || !(end * 1000.0).is_finite()
            {
                return Err(failure("Cartesia returned an invalid timestamp"));
            }
            Ok(out::Timestamp {
                kind: if phoneme {
                    out::TimestampKind::Phoneme(Default::default())
                } else {
                    out::TimestampKind::Word(Default::default())
                },
                value,
                start_time_ms: start * 1000.0,
                end_time_ms: end * 1000.0,
            })
        })
        .collect()
}

pub(super) fn output(
    packet: Packet,
    id: &str,
    timed: bool,
) -> Result<Option<out::SynthesisItem>, TransportError> {
    let (context, audio, timestamps) = match packet {
        Packet::Failure(_, error) => return Err(Box::new(error)),
        Packet::Done(_) => return Ok(None),
        Packet::Flushed(_, group) => {
            return Ok(Some(out::SynthesisItem::Flush(out::SynthesisItemFlush {
                event: Default::default(),
                correlation_id: id.into(),
                input_group_id: Some(group),
            })))
        }
        Packet::Chunk(context, data) => {
            let audio = base64::decode(&data)
                .ok_or_else(|| failure("Cartesia returned invalid base64 audio"))?;
            if !timed && context.group.is_none() {
                return Ok(Some(out::SynthesisItem::Bytes(audio)));
            }
            (context, Some(audio), Vec::new())
        }
        Packet::Timing(context, timestamps) => (context, None, timestamps),
    };
    Ok(Some(out::SynthesisItem::Timeline(out::TimelineOutput {
        correlation: Default::default(),
        correlation_id: id.into(),
        input_group_id: context.group,
        audio,
        timestamps,
    })))
}
